kubetest2/pkg/app/shim/deployers.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shim
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// FindDeployer locates the binary implementing the named deployer
+// TODO(bentheelder): move this to another package?
+func FindDeployer(name string) (path string, err error) {
+	binary := fmt.Sprintf("%s-%s", BinaryName, name)
+	path, err = exec.LookPath(binary)
+	if err != nil {
+		return "", errors.Errorf("%#v not found in PATH, could not locate %#v deployer", binary, name)
+	}
+	return path, err
+}
+
+// FindDeployers looks for all deployers in PATH, returning a map of the
+// deployer name to the first matching binary found in path
+func FindDeployers() map[string]string {
+	nameToPath := make(map[string]string)
+	prefix := fmt.Sprintf("%s-", BinaryName)
+
+	// search every directory in PATH for kubetest2-* binaries
+	searchPaths := filepath.SplitList(os.Getenv("PATH"))
+	for _, dir := range searchPaths {
+		// mimic LookPath() for nicer results
+		if dir == "" {
+			// Unix shell semantics: path element "" means "."
+			dir = "."
+		}
+
+		// list all files in the directory
+		files, err := ioutil.ReadDir(dir)
+
+		// ignore bad directories in PATH
+		if _, ok := err.(*os.PathError); ok && strings.Contains(err.Error(), "no such file") {
```

Comment:
hummm... use [Stat](https://golang.org/pkg/os/#Stat)? this error check looks hacky

Reply:
OK, FWIW this error check is actually borrowed from kubectl's plugin logic, which I read before implementing the shim to make sure we weren't re-inventing the world unnecessarily (unfortunately their logic is pretty coupled and not re-usable).

Reply:
actually even cleaner:
```go
// ignore bad directories in PATH
if os.IsNotExist(err) {
	continue
}
```

Reply:
tested with `PATH="/a:$PATH" go run ./kubetest2/main.go -h` and a `println` inside this if :-)

Reply:
(https://github.com/kubernetes/kubernetes/pull/73542 🤷‍♂️anyway)